nimpy.nim

```diff
@@ -957,6 +1016,142 @@ proc pyObjToNimSeq[T](o: PPyObject, v: var seq[T]) =
         pyObjToNim(pyLib.PyList_GetItem(o, i), v[i])
         # PyList_GetItem # No DECREF. Returns borrowed ref.
 
+
+proc pyObjToNimTab[T; U](o: PPyObject, tab: var Table[T, U]) =
+    ## call this either:
+    ## - if you want to check whether T and U are valid types for
+    ##   the python dict (i.e. to check whether all python types
+    ##   are convertible to T and U)
+    ## - you know the python dict conforms to T and U and you wish
+    ##   to get a correct Nim table from that
+    tab = initTable[T, U]()
+    let
+        sz = int(pyLib.PyDict_Size(o))
+        ks = pyLib.PyDict_Keys(o)
+        vs = pyLib.PyDict_Values(o)
+    for i in 0 ..< sz:
+        var
+            k: T
+            v: U
+        pyObjToNim(pyLib.PyList_GetItem(ks, i), k)
+        pyObjToNim(pyLib.PyList_GetItem(vs, i), v)
+        # PyList_GetItem # No DECREF. Returns borrowed ref.
+        tab[k] = v
+    decref ks
+    decref vs
+
+proc `%`(o: PPyObject): JsonNode
+proc `%`(o: PyObject): JsonNode
+
+template getVal(pyVal: PPyObject, actions: untyped) {.dirty.} =
+    # dirty template to reduce boilerplate. `checkPyObjPyType` determines
+    # the type of a `PPyObject` using the
+    let kType = checkPyObjPyType(pyVal)
+    case kType
+    of pbLong:
+        var x: int
+        pyObjToNim(pyVal, x)
+        block:
+            actions
+    of pbFloat:
+        var x: float
+        pyObjToNim(pyVal, x)
+        block:
+            actions
+    of pbComplex:
+        var x: Complex
+        pyObjToNim(pyVal, x)
+        block:
+            actions
+    of pbDict:
+        var x = newJObject()
+        pyObjToNim(pyVal, x)
+        block:
+            actions
+    of pbBytes, pbUnicode, pbString:
+        var x = ""
+        pyObjToNim(pyVal, x)
+        block:
+            actions
+    of pbList:
+        # using % converting python list to Json JArray
+        var x = % pyVal
+        block:
+            actions
+    of pbTuple:
+        # converts to JArray as well
+        var x = % pyVal
+        block:
+            actions
+    else:
+        # for now we simply convert objects to their string
+        # representation. More useful than either crashing or
+        # dropping quietly
+        var x = % newPyObject(pyVal)
```

Comment:
Can't you use `%` directly on `pyVal` here?

Reply:
With the current implementation, no (which is probably valid criticism), because I steal the `%` for `PPyObject` for Python dictionaries.